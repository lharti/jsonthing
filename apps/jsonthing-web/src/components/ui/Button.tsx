import { cn } from '@/lib/utils'
import { Slot } from '@radix-ui/react-slot'
import { cva, type VariantProps } from 'class-variance-authority'
import * as React from 'react'

const buttonVariants = cva(
    `
      inline-flex items-center justify-center whitespace-nowrap rounded-md
      text-sm font-medium transition-colors

      dark:focus-visible:ring-gray-300

      disabled:pointer-events-none disabled:opacity-50

      focus-visible:outline-none focus-visible:ring-1
      focus-visible:ring-gray-950
    `,
    {
        variants: {
            variant: {
                default: `
                  bg-gray-900 text-gray-50 shadow

                  dark:bg-gray-50 dark:text-gray-900 dark:hover:bg-gray-50/90

                  hover:bg-gray-900/90
                `,
                destructive: `
                  bg-red-500 text-gray-50 shadow-sm

                  dark:bg-red-900 dark:text-gray-50 dark:hover:bg-red-900/90

                  hover:bg-red-500/90
                `,
                outline: `
                  border border-gray-200 bg-white shadow-sm

                  dark:border-gray-800 dark:bg-gray-950 dark:hover:bg-gray-800
                  dark:hover:text-gray-50

                  hover:bg-gray-100 hover:text-gray-900
                `,
                secondary: `
                  bg-gray-100 text-gray-900 shadow-sm

                  dark:bg-gray-800 dark:text-gray-50 dark:hover:bg-gray-800/80

                  hover:bg-gray-100/80
                `,
                ghost: `
                  dark:hover:bg-gray-800 dark:hover:text-gray-50

                  hover:bg-gray-100 hover:text-gray-900
                `,
                link: `
                  text-gray-900 underline-offset-4

                  dark:text-gray-50

                  hover:underline
                `,
            },
            size: {
                default: 'h-9 px-4 py-2',
                sm: 'h-8 rounded-md px-3 text-xs',
                lg: 'h-10 rounded-md px-8',
                icon: 'size-9',
            },
        },
        defaultVariants: {
            variant: 'default',
            size: 'default',
        },
    },
)

export interface ButtonProps
    extends React.ButtonHTMLAttributes<HTMLButtonElement>,
        VariantProps<typeof buttonVariants> {
    readonly asChild?: boolean
}

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
    ({ className, variant, size, asChild = false, ...props }, ref) => {
        const Comp = asChild ? Slot : 'button'
        return (
            <Comp
                className={cn(buttonVariants({ variant, size, className }))}
                ref={ref}
                {...props}
            />
        )
    },
)

Button.displayName = 'Button'

export { Button, buttonVariants }
