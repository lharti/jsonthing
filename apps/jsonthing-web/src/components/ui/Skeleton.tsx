import { cn } from '@/lib/utils'
import React from 'react'

const Skeleton = ({
    className,
    ...props
}: React.HTMLAttributes<HTMLDivElement>) => {
    return (
        <div
            className={cn(
                `
                  animate-pulse rounded-md bg-gray-900/10

                  dark:bg-gray-50/10
                `,
                className,
            )}
            // eslint-disable-next-line react/jsx-props-no-spreading
            {...props}
        />
    )
}

export { Skeleton }
