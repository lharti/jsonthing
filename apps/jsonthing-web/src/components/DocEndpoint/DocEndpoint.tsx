import { Button } from '@/components/ui/Button'
import {
    Tooltip,
    TooltipContent,
    TooltipTrigger,
} from '@/components/ui/Tooltip'
import { cn } from '@/lib/utils'
import { IconApi, IconClipboardCopy } from '@tabler/icons-react'
import Link from 'next/link'
import React from 'react'

export interface DocEndpointProps {
    docId: string
    className?: string
}

export const DocEndpoint: React.FC<DocEndpointProps> = ({
    docId,
    className,
}) => {
    const docEndpointUrl = `${process.env.NEXT_PUBLIC_API_URL}/docs/${docId}`

    return (
        <div className={cn('flex items-center', className)}>
            <p className="mr-3 rounded-sm bg-green-500 p-1 text-white">
                <IconApi />
            </p>

            <Link
                href={docEndpointUrl}
                target="_blank"
                className={`
                  mr-3 overflow-hidden text-ellipsis text-sm font-bold
                  text-gray-400
                `}
            >
                {docEndpointUrl}
            </Link>

            <Tooltip>
                <TooltipTrigger asChild>
                    <Button
                        variant="ghost"
                        size="sm"
                        role="button"
                        aria-label="Copy endpoint URL"
                        className="text-gray-600"
                        onClick={() => {
                            navigator.clipboard.writeText(docEndpointUrl)
                        }}
                    >
                        <IconClipboardCopy />
                    </Button>
                </TooltipTrigger>

                <TooltipContent side="bottom">
                    {'Copy endpoint URL'}
                </TooltipContent>
            </Tooltip>
        </div>
    )
}
