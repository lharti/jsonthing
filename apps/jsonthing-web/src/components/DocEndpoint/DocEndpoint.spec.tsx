import { TooltipProvider } from '@/components/ui/Tooltip'
import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React, { act } from 'react'
import { DocEndpoint } from './DocEndpoint'

describe('<DocEndpoint />', () => {
    it('should render', () => {
        expect.assertions(1)

        process.env.NEXT_PUBLIC_API_URL = 'API_URL'

        const { container } = render(<DocEndpoint docId="DOC_ID" />, {
            wrapper: TooltipProvider,
        })

        expect(container).toMatchSnapshot()
    })

    it('should copy endpoint URL to clipboard', async () => {
        expect.assertions(1)

        userEvent.setup()

        process.env.NEXT_PUBLIC_API_URL = 'API_URL'

        const { getByRole } = render(<DocEndpoint docId="DOC_ID" />, {
            wrapper: TooltipProvider,
        })

        const copyButton = getByRole('button', { name: 'Copy endpoint URL' })

        await act(async () => {
            await copyButton.click()
        })

        const clipboardContent = await navigator.clipboard.readText()

        expect(clipboardContent).toMatchInlineSnapshot(`"API_URL/dc/DOC_ID"`)
    })
})
