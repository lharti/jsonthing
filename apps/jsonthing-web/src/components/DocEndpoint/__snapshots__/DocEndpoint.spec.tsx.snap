// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DocEndpoint /> should render 1`] = `
<div>
  <div
    class="flex items-center"
  >
    <p
      class="mr-3 rounded-sm bg-green-500 p-1 text-white"
    >
      <svg
        class="tabler-icon tabler-icon-api "
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 13h5"
        />
        <path
          d="M12 16v-8h3a2 2 0 0 1 2 2v1a2 2 0 0 1 -2 2h-3"
        />
        <path
          d="M20 8v8"
        />
        <path
          d="M9 16v-5.5a2.5 2.5 0 0 0 -5 0v5.5"
        />
      </svg>
    </p>
    <a
      class="
                  mr-3 overflow-hidden text-ellipsis text-sm font-bold
                  text-gray-400
                "
      href="API_URL/dc/DOC_ID"
      target="_blank"
    >
      API_URL/dc/DOC_ID
    </a>
    <button
      aria-label="Copy endpoint URL"
      class="inline-flex items-center justify-center whitespace-nowrap font-medium transition-colors dark:focus-visible:ring-gray-300 disabled:pointer-events-none disabled:opacity-50 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-950 dark:hover:bg-gray-800 dark:hover:text-gray-50 hover:bg-gray-100 hover:text-gray-900 h-8 rounded-md px-3 text-xs text-gray-600"
      data-state="closed"
      role="button"
    >
      <svg
        class="tabler-icon tabler-icon-clipboard-copy "
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9 5h-2a2 2 0 0 0 -2 2v12a2 2 0 0 0 2 2h3m9 -9v-5a2 2 0 0 0 -2 -2h-2"
        />
        <path
          d="M13 17v-1a1 1 0 0 1 1 -1h1m3 0h1a1 1 0 0 1 1 1v1m0 3v1a1 1 0 0 1 -1 1h-1m-3 0h-1a1 1 0 0 1 -1 -1v-1"
        />
        <path
          d="M9 3m0 2a2 2 0 0 1 2 -2h2a2 2 0 0 1 2 2v0a2 2 0 0 1 -2 2h-2a2 2 0 0 1 -2 -2z"
        />
      </svg>
    </button>
  </div>
</div>
`;
